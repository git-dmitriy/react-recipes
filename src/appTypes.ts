export type CategoryItemTypes = {
  idCategory: string;
  strCategory: string;
  strCategoryThumb: string;
  strCategoryDescription: string;
};

export type MealItemTypes = {
  idMeal: string;
  strMeal: string;
  strMealThumb: string;
  strCategory?: string;
  strArea?: string;
  strInstructions?: string;
  strYoutube?: string;
};

export type RecipeItemTypes = {
  idMeal: string;
  strMeal: string;
  strMealThumb: string;
};

export type StateTypes = {
  favorites: MealItemTypes[] | [];
  theme: string | 'light' | 'dark';
  isLoading: boolean;
};

export type ActionTypes =
  | { type: ADD_TO_FAVORITE; payload: MealItemTypes }
  | { type: REMOVE_FROM_FAVORITE; payload: MealItemTypes }
  | { type: SWITCH_THEME; payload: 'light' | 'dark' }
  | { type: SET_IS_LOADING; payload: boolean };

type ADD_TO_FAVORITE = 'ADD_TO_FAVORITE';
type REMOVE_FROM_FAVORITE = 'REMOVE_FROM_FAVORITE';
type SWITCH_THEME = 'SWITCH_THEME';
type SET_IS_LOADING = 'SET_IS_LOADING';

export type ContextTypes = {
  state: StateTypes;
  addToFavorites: (item: MealItemTypes) => void;
  removeFromFavorites: (item: MealItemTypes) => void;
  switchTheme: (mode: 'light' | 'dark') => void;
  setIsLoading: (status: boolean) => void;
};
