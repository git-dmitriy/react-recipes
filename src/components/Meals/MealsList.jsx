import { Meal } from './Meal';

export function MealsList({ meals }) {
  return (
    <>
      <ul className='mx-auto grid xl:grid-cols-3 md:grid-cols-2 grid-cols-1 gap-2 xl:max-w-6xl max-w-4xl'>
        {meals.map((meal) => (
          <Meal key={meal.idMeal} {...meal} />
        ))}
      </ul>
    </>
  );
}
