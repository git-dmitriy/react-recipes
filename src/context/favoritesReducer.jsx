import { ADD_TO_FAVORITE, REMOVE_FROM_FAVORITE } from './types';

const handlers = {
  [ADD_TO_FAVORITE]: (state, { payload }) => {
    return {
      favorites: [...state.favorites, payload],
    };
  },
  [REMOVE_FROM_FAVORITE]: (state, { payload }) => ({
    favorites: state.favorites.filter((item) => item.idMeal !== payload.idMeal),
  }),
  DEFAULT: (state) => state,
};

export const favoriteReducer = (state, action) => {
  const handler = handlers[action.type] || handlers.DEFAULT;
  return handler(state, action);
};
