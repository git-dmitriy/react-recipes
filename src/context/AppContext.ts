import { createContext } from 'react';
import { ContextTypes } from '@/appTypes';

export const AppContext = createContext({} as ContextTypes);
