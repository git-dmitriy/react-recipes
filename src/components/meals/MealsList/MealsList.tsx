import { Meal } from 'components/meals/Meals';
import { MealItemTypes } from 'appTypes';
import { motion } from 'framer-motion';

type P = {
  meals: MealItemTypes[];
};

export const MealsList: React.FC<P> = ({ meals }) => {
  const variants = {
    visible: (item: number) => ({
      opacity: 1,
      transition: {
        delay: item * 0.03,
      },
    }),
    hidden: { opacity: 0 },
  };

  return (
    <ul className='mx-auto grid grid-cols-1 sm:grid-cols-2 xl:grid-cols-3 gap-10 justify-items-center xl:max-w-6xl max-w-4xl'>
      {meals.map((meal, idx) => (
        <motion.li
          key={meal.idMeal}
          className='flex'
          variants={variants}
          initial='hidden'
          animate='visible'
          custom={idx}
        >
          <Meal {...meal} />
        </motion.li>
      ))}
    </ul>
  );
};
