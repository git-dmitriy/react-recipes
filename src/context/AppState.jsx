import React, { useReducer } from 'react';
import {
  ADD_TO_FAVORITE,
  SWITCH_THEME,
  REMOVE_FROM_FAVORITE,
} from 'context/types';
import { AppContext } from 'context/AppContext';
import { reducer } from 'context/reducer';

export const AppState = ({ children }) => {
  const favorites = window.localStorage.getItem('favorites');
  const initialState = {
    favorites: JSON.parse(favorites) || [],
    theme: window.localStorage.getItem('theme') || 'light',
  };

  const [state, dispatch] = useReducer(reducer, initialState);

  const addToFavorites = (item) => {
    dispatch({
      type: ADD_TO_FAVORITE,
      payload: item,
    });
  };

  const removeFromFavorites = (item) => {
    dispatch({
      type: REMOVE_FROM_FAVORITE,
      payload: item,
    });
  };

  const switchTheme = (mode) => {
    if (mode === 'light' || mode === 'dark') {
      dispatch({
        type: SWITCH_THEME,
        payload: mode,
      });
    }
  };

  return (
    <AppContext.Provider
      value={{
        addToFavorites,
        removeFromFavorites,
        switchTheme,
        state,
      }}
    >
      {children}
    </AppContext.Provider>
  );
};
