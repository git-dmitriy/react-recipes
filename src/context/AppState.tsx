import React, { useReducer } from 'react';
import {
  ADD_TO_FAVORITE,
  SWITCH_THEME,
  REMOVE_FROM_FAVORITE,
  SET_IS_LOADING,
} from 'context/types';
import { AppContext } from 'context/AppContext';
import { reducer } from 'context/reducer';
import { StateTypes, MealItemTypes, ContextTypes } from 'appTypes';

type P = { children: React.ReactNode };

export const AppState = ({ children }: P) => {
  const favorites = window.localStorage.getItem('favorites') || '[]';

  const initialState: StateTypes = {
    favorites: JSON.parse(favorites),
    theme: window.localStorage.getItem('theme') || 'light',
    isLoading: false,
  };

  const [state, dispatch] = useReducer(reducer, initialState);

  const addToFavorites = (item: MealItemTypes) => {
    dispatch({
      type: ADD_TO_FAVORITE,
      payload: item,
    });
  };

  const removeFromFavorites = (item: MealItemTypes) => {
    dispatch({
      type: REMOVE_FROM_FAVORITE,
      payload: item,
    });
  };

  const switchTheme = (mode: string) => {
    if (mode === 'light' || mode === 'dark') {
      dispatch({
        type: SWITCH_THEME,
        payload: mode,
      });
    }
  };

  const setIsLoading = (status = false) => {
    dispatch({
      type: SET_IS_LOADING,
      payload: status,
    });
  };

  const value: ContextTypes = {
    addToFavorites,
    removeFromFavorites,
    switchTheme,
    setIsLoading,
    state,
  };

  return <AppContext.Provider value={value}>{children}</AppContext.Provider>;
};
