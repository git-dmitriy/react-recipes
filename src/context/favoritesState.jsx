import React, { useReducer } from 'react';
import { ADD_TO_FAVORITE, REMOVE_FROM_FAVORITE } from 'context/types';
import { FavoriteContext } from 'context/favoritesContext';
import { favoriteReducer } from 'context/favoritesReducer';

export const FavoritesState = ({ children }) => {
  const favorites = window.localStorage.getItem('favorites');
  const initialState = {
    favorites: JSON.parse(favorites) || [],
  };

  const [state, dispatch] = useReducer(favoriteReducer, initialState);

  const addToFavorites = (item) => {
    dispatch({
      type: ADD_TO_FAVORITE,
      payload: item,
    });
  };

  const removeFromFavorites = (item) => {
    dispatch({
      type: REMOVE_FROM_FAVORITE,
      payload: item,
    });
  };

  return (
    <FavoriteContext.Provider
      value={{
        addToFavorites,
        removeFromFavorites,
        state,
      }}
    >
      {children}
    </FavoriteContext.Provider>
  );
};
