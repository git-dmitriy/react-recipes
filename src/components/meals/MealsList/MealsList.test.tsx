import renderer from 'react-test-renderer';
import { MealsList } from './MealsList';
import { MemoryRouter } from 'react-router-dom';
import { AppContext } from 'context/AppContext';
import { ContextTypes } from 'appTypes';

const context = {
  state: {
    theme: 'light',
  },
};

const data = [
  {
    idMeal: '52965',
    strMeal: 'Breakfast Potatoes',
    strMealThumb: 'https://www.themealdb.com/images/media/meals/1550441882.jpg',
  },
];

it('should render correctly', () => {
  const tree = renderer
    .create(
      <AppContext.Provider value={context as ContextTypes}>
        <MemoryRouter>
          <MealsList meals={data} />
        </MemoryRouter>
      </AppContext.Provider>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
