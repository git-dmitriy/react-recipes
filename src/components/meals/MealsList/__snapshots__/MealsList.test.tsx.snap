// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<ul
  className="mx-auto grid grid-cols-1 sm:grid-cols-2 xl:grid-cols-3 gap-10 justify-items-center xl:max-w-6xl max-w-4xl"
>
  <li
    className="flex"
    style={
      Object {
        "opacity": 0,
      }
    }
  >
    <article
      className="relative w-full z-0 bg-white dark:bg-gray-100 rounded-3xl overflow-hidden border border-gray-200 dark:border-gray-900 dark:text-gray-900"
    >
      <a
        className="block rounded-lg overflow-hidden min-h-300 hover:opacity-90 transition-opacity"
        href="/meal/52965"
        onClick={[Function]}
      >
        <div
          className="lazyload-wrapper "
        >
          <div
            className="lazyload-placeholder"
            style={
              Object {
                "height": 437,
              }
            }
          />
        </div>
        <div
          className="text-center my-2 mx-3 min-h-full"
        >
          <h4
            className="text-2xl font-semibold"
          >
            Breakfast Potatoes
          </h4>
        </div>
      </a>
      <div
        className="absolute top-4 right-4 p-1 bg-white bg-opacity-50 backdrop-opacity-50 backdrop-filter backdrop-blur-2xl rounded-2xl"
      >
        <div
          className="mx-2 mt-2"
        >
          <button
            className="text-2xl text-gray-900  hover:text-red-500 transition duration-200 uppercase"
            onClick={[Function]}
          >
            <svg
              className="fill-current"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M2 15.5V2a2 2 0 0 1 2-2h8a2 2 0 0 1 2 2v13.5a.5.5 0 0 1-.74.439L8 13.069l-5.26 2.87A.5.5 0 0 1 2 15.5zm6.5-11a.5.5 0 0 0-1 0V6H6a.5.5 0 0 0 0 1h1.5v1.5a.5.5 0 0 0 1 0V7H10a.5.5 0 0 0 0-1H8.5V4.5z"
                fillRule="evenodd"
              />
            </svg>
          </button>
        </div>
      </div>
    </article>
  </li>
</ul>
`;
